repose-aggregator/components/filters/openstack-identity-v3-filter/src/main/scala/org/openrepose/filters/openstackidentityv3/OpenStackIdentityV3Filter.scala

```diff
@@ -87,17 +96,81 @@ class OpenStackIdentityV3Filter @Inject()(configurationService: ConfigurationSer
   def isInitialized = initialized
 
   override def destroy() {
-    Option(akkaServiceClient).foreach(_.destroy())
     configurationService.unsubscribeFrom(configFilename, this)
+    CacheInvalidationFeedListener.unregisterFeeds()
+    Option(akkaServiceClient).foreach(_.destroy())
   }
 
   def configurationUpdated(config: OpenstackIdentityV3Config) {
+    // This will also un-register any Atom Feeds not present in the new config.
+    CacheInvalidationFeedListener.updateFeeds(config.getAtomFeedIds.toSet)
+
     val akkaServiceClientOld = Option(akkaServiceClient)
     akkaServiceClient = akkaServiceClientFactory.newAkkaServiceClient(config.getConnectionPoolId)
     akkaServiceClientOld.foreach(_.destroy())
 
-    val identityAPI = new OpenStackIdentityV3API(config, datastoreService.getDefaultDatastore, akkaServiceClient)
+    val identityAPI = new OpenStackIdentityV3API(config, cache, akkaServiceClient)
     openStackIdentityV3Handler = new OpenStackIdentityV3Handler(config, identityAPI)
     initialized = true
   }
+
+  object CacheInvalidationFeedListener extends AtomFeedListener {
+
+    private var registeredFeeds = List.empty[RegisteredFeed]
+
+    def unregisterFeeds(): Unit = {
+      registeredFeeds synchronized {
+        registeredFeeds foreach { feed =>
+          atomFeedService.unregisterListener(feed.unique)
+        }
+      }
+    }
+
+    def updateFeeds(feeds: Set[String]): Unit = {
+      registeredFeeds synchronized {
+        // Unregister the feeds we no longer care about.
+        val feedsToUnregister = registeredFeeds.filterNot(regFeed => feeds.contains(regFeed.id))
+        feedsToUnregister.foreach(feed => atomFeedService.unregisterListener(feed.unique))
+
+        // Register with only the new feeds we aren't already registered with.
+        val registeredFeedIds = registeredFeeds.map(_.id)
+        val feedsToRegister = feeds.filterNot(posFeed => registeredFeedIds.contains(posFeed))
+        val newRegisteredFeeds = feedsToRegister.map(feedId => RegisteredFeed(feedId, atomFeedService.registerListener(feedId, this)))
+
+        // Update to the still and newly registered feeds.
+        registeredFeeds = registeredFeeds.diff(feedsToUnregister) ++ newRegisteredFeeds
+      }
+    }
+
+    override def onNewAtomEntry(atomEntry: String): Unit = {
+      logger.debug("Processing atom feed entry: {}", atomEntry)
+      val atomXml = XML.loadString(atomEntry)
+      val resourceId = (atomXml \\ "event" \\ "@resourceId").map(_.text).headOption
+      if (resourceId.isDefined) {
+        val resourceType = (atomXml \\ "event" \\ "@resourceType").map(_.text)
+        val authTokens = resourceType.headOption match {
+          // User OR Token Revocation Record (TRR) event.
+          case Some("USER") | Some("TRR_USER") =>
+            val tokens = Option(cache.get(getUserIdKey(resourceId.get)).asInstanceOf[PatchableSet[String]])
```

Comment:
I kind of have to assume you are otherwise this wouldn't work, but where and why?